internal/host/host.go

```diff
@@ -1412,3 +1418,99 @@ func (m *Manager) HandleReclaimBootArtifactDownload(ctx context.Context, h *mode
 func (m *Manager) HandleReclaimFailure(ctx context.Context, h *models.Host) error {
 	return m.sm.Run(TransitionTypeReclaimFailed, newStateHost(h), &TransitionArgsUnbindHost{ctx: ctx, db: m.db})
 }
+
+// autorenameHost renames a host if it is necessary and possible. For example, if the name is
+// localhost.localdomain and the MAC address of the main network interface is A8:CD:16:AE:79:01 it
+// will be renamed to a8-cd-16-ae-79-01. Returns a boolean indicating if the host was actually
+// renamed.
+func (m *Manager) autorenameHost(host *models.Host, inventory *models.Inventory) bool {
+	if host.RequestedHostname != "" {
+		return false
+	}
+	if inventory == nil || inventory.Hostname == "" {
+		return false
+	}
+	if !funk.Contains(hostutil.ForbiddenHostnames, inventory.Hostname) {
+		return false
+	}
+	if len(inventory.Interfaces) < 1 {
+		return false
+	}
+	mainNIC, err := m.mainNIC(inventory)
+	if err != nil {
+		m.log.WithError(err).WithFields(logrus.Fields{
+			"host": host.ID,
+			"name": inventory.Hostname,
+		}).Error(
+			"Host hasn't been automatically renamed because an unexpected error " +
+				"ocurred while trying to find the main network interface card",
+		)
+		return false
+	}
+	if mainNIC == nil {
+		m.log.WithFields(logrus.Fields{
+			"host": host.ID,
+			"name": inventory.Hostname,
+		}).Info(
+			"Host hasn't been automatically renamed because there is no network " +
+				"interface card with a MAC address that can be used to calculate " +
+				"the new name",
+		)
+		return false
+	}
+	host.RequestedHostname = strings.ToLower(strings.ReplaceAll(mainNIC.MacAddress, ":", "-"))
+	m.log.WithFields(logrus.Fields{
+		"host": host.ID,
+		"old":  inventory.Hostname,
+		"new":  host.RequestedHostname,
+		"nic":  mainNIC.Name,
+		"mac":  mainNIC.MacAddress,
+	}).Info("Host has been automatically renamed")
+	return true
+}
+
+// mainNIC returns the main network interface of the given host inventory, or nil if there is no
+// such network interface card. The main network interface card is the first that has assigned both
+// a MAC address and a global IP address.
+func (m *Manager) mainNIC(inventory *models.Inventory) (result *models.Interface, err error) {
```

Comment:
Also a naming nit: I have a hard time with digesting the concept of using the name "main network interface" for an interface that just have a MAC and IP address assigned. I definitely like the logic you do here so that you will pick an usable NIC and build a hostname out of it. It's just that in a global perspective "main network interface" may have another meaning.

I wonder if something as simple as `findUsableNIC` may satisfy what you are trying to do here? Given that the result is not guaranteed to be deterministic as a function of the host (because the content of `inventory.Interfaces` has no guaranteed order) but also given that we effectively don't care which NIC is used for this autorenaming, this could be some tradeoff.

If you really really wanted it to be about detecting the main network interface, then I'd suggest not picking "the first" interface but rather the interface that matches the Machine Networks of the cluster. But this path requires you to parse the whole Cluster object here and cross-check it with host's inventory. So looks like a big overkill for what you try to achieve

Reply:
Renamed to `findUsableNIC`. I agree that cross checking with the cluster networks isn't required here.